File: diagram_agent/tools.py
"""
Diagram Agent Tools

Reusable tools for diagram validation, rendering, and file operations.
"""

import subprocess
import tempfile
import os
import re
import base64
import urllib.request
import zlib
from typing import Optional, Tuple, List
from dataclasses import dataclass
from enum import Enum


class DiagramFormat(Enum):
    GRAPHVIZ = "graphviz"
    PLANTUML = "plantuml"
    MERMAID = "mermaid"


@dataclass
class ValidationResult:
    """Result of diagram validation."""
    valid: bool
    format: DiagramFormat
    errors: List[str]
    code: str
    suggestions: List[str] = None

    def __post_init__(self):
        if self.suggestions is None:
            self.suggestions = []


@dataclass
class RenderResult:
    """Result of diagram rendering."""
    success: bool
    output_path: Optional[str]
    error: Optional[str]
    format: DiagramFormat


# ============== Validation Tools ==============

def validate_graphviz(code: str) -> ValidationResult:
    """
    Validate Graphviz DOT code.

    Args:
        code: The DOT code to validate

    Returns:
        ValidationResult with validation details
    """
    errors = []
    suggestions = []

    # Basic syntax checks
    if not re.search(r'(di)?graph\s+\w+\s*\{', code, re.IGNORECASE):
        errors.append("No valid graph definition found (digraph Name { or graph Name {)")
        suggestions.append("Start with: digraph MyDiagram {")

    # Check brace balance
    open_braces = code.count('{')
    close_braces = code.count('}')
    if open_braces != close_braces:
        errors.append(f"Unbalanced braces: {open_braces} '{{' vs {close_braces} '}}'")
        suggestions.append("Ensure every '{' has a matching '}'")

    # Check for common LLM mistakes
    if re.search(r'\.\w+\s*\[', code):
        errors.append("Node names cannot start with dots (e.g., .env is invalid)")
        suggestions.append("Use underscores: env_file instead of .env")

    if re.search(r':\w+\s*->', code):
        errors.append("Port syntax (:port) in node names may cause issues")
        suggestions.append("Define ports separately using record shapes")

    # If basic checks pass, try actual validation with dot
    if not errors:
        try:
            with tempfile.NamedTemporaryFile(mode='w', suffix='.dot', delete=False) as f:
                f.write(code)
                temp_path = f.name

            result = subprocess.run(
                ["dot", "-Tcanon", temp_path],
                capture_output=True,
                text=True,
                timeout=10
            )

            os.unlink(temp_path)

            if result.returncode != 0:
                for line in result.stderr.strip().split('\n'):
                    if line.strip():
                        errors.append(line.strip())

        except FileNotFoundError:
            errors.append("Graphviz not installed. Install with: brew install graphviz")
        except subprocess.TimeoutExpired:
            errors.append("Validation timed out - code may be too complex")
        except Exception as e:
            errors.append(f"Validation error: {str(e)}")

    return ValidationResult(
        valid=len(errors) == 0,
        format=DiagramFormat.GRAPHVIZ,
        errors=errors,
        code=code,
        suggestions=suggestions
    )


def validate_mermaid(code: str) -> ValidationResult:
    """
    Validate Mermaid diagram code.

    Args:
        code: The Mermaid code to validate

    Returns:
        ValidationResult with validation details
    """
    errors = []
    suggestions = []

    code_stripped = code.strip()

    # Check for valid start
    valid_starts = ['flowchart', 'graph', 'sequencediagram', 'classdiagram',
                   'erdiagram', 'gantt', 'pie', 'statediagram']

    has_valid_start = any(code_stripped.lower().startswith(s) for s in valid_starts)
    if not has_valid_start:
        errors.append("Must start with a valid diagram type (flowchart, graph, etc.)")
        suggestions.append("Start with: flowchart LR or graph TD")

    # Check bracket balance
    if code.count('[') != code.count(']'):
        errors.append(f"Unbalanced square brackets: {code.count('[')} '[' vs {code.count(']')} ']'")
        suggestions.append("Ensure every '[' has a matching ']'")

    if code.count('(') != code.count(')'):
        errors.append(f"Unbalanced parentheses: {code.count('(')} '(' vs {code.count(')')} ')'")

    # Check subgraph/end balance
    subgraph_count = len(re.findall(r'\bsubgraph\b', code, re.IGNORECASE))
    end_count = len(re.findall(r'\bend\b', code, re.IGNORECASE))
    if subgraph_count != end_count:
        errors.append(f"Unbalanced subgraph/end: {subgraph_count} subgraph vs {end_count} end")
        suggestions.append("Every 'subgraph' needs a closing 'end'")

    return ValidationResult(
        valid=len(errors) == 0,
        format=DiagramFormat.MERMAID,
        errors=errors,
        code=code,
        suggestions=suggestions
    )


def validate_plantuml(code: str) -> ValidationResult:
    """
    Validate PlantUML diagram code.

    Args:
        code: The PlantUML code to validate

    Returns:
        ValidationResult with validation details
    """
    errors = []
    suggestions = []

    # Check for @startuml/@enduml
    if '@startuml' not in code.lower():
        errors.append("Missing @startuml at the beginning")
        suggestions.append("Start with: @startuml")

    if '@enduml' not in code.lower():
        errors.append("Missing @enduml at the end")
        suggestions.append("End with: @enduml")

    # Check order
    if '@startuml' in code.lower() and '@enduml' in code.lower():
        start_pos = code.lower().find('@startuml')
        end_pos = code.lower().find('@enduml')
        if start_pos > end_pos:
            errors.append("@startuml must come before @enduml")

    return ValidationResult(
        valid=len(errors) == 0,
        format=DiagramFormat.PLANTUML,
        errors=errors,
        code=code,
        suggestions=suggestions
    )


def validate_diagram(code: str, format_hint: str = None) -> ValidationResult:
    """
    Validate diagram code, auto-detecting format if not specified.

    Args:
        code: The diagram code to validate
        format_hint: Optional format hint ("graphviz", "mermaid", "plantuml")

    Returns:
        ValidationResult with validation details
    """
    code = code.strip()

    # Auto-detect format
    if format_hint:
        detected_format = format_hint.lower()
    elif '@startuml' in code.lower():
        detected_format = 'plantuml'
    elif code.lower().startswith(('flowchart', 'graph ', 'sequencediagram')):
        detected_format = 'mermaid'
    elif re.search(r'(di)?graph\s+\w+\s*\{', code, re.IGNORECASE):
        detected_format = 'graphviz'
    else:
        # Default to graphviz
        detected_format = 'graphviz'

    validators = {
        'graphviz': validate_graphviz,
        'plantuml': validate_plantuml,
        'mermaid': validate_mermaid,
    }

    return validators.get(detected_format, validate_graphviz)(code)


# ============== Rendering Tools ==============

def render_graphviz(code: str, output_path: str) -> RenderResult:
    """Render Graphviz DOT code to PNG."""
    try:
        with tempfile.NamedTemporaryFile(mode='w', suffix='.dot', delete=False) as f:
            f.write(code)
            temp_path = f.name

        result = subprocess.run(
            ["dot", "-Tpng", "-Gdpi=150", temp_path, "-o", output_path],
            capture_output=True,
            text=True,
            timeout=30
        )

        os.unlink(temp_path)

        if result.returncode == 0:
            return RenderResult(True, output_path, None, DiagramFormat.GRAPHVIZ)
        else:
            return RenderResult(False, None, result.stderr, DiagramFormat.GRAPHVIZ)

    except Exception as e:
        return RenderResult(False, None, str(e), DiagramFormat.GRAPHVIZ)


def render_mermaid(code: str, output_path: str) -> RenderResult:
    """Render Mermaid code to PNG using web service."""
    try:
        encoded = base64.urlsafe_b64encode(code.encode('utf-8')).decode('utf-8')
        url = f"https://mermaid.ink/img/{encoded}?type=png&bgColor=white"

        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        response = urllib.request.urlopen(req, timeout=30)

        with open(output_path, 'wb') as f:
            f.write(response.read())

        return RenderResult(True, output_path, None, DiagramFormat.MERMAID)

    except Exception as e:
        return RenderResult(False, None, str(e), DiagramFormat.MERMAID)


def render_plantuml(code: str, output_path: str) -> RenderResult:
    """Render PlantUML code to PNG using web service."""
    def encode_plantuml(text):
        compressed = zlib.compress(text.encode('utf-8'))[2:-4]
        res = ""
        for i in range(0, len(compressed), 3):
            if i + 2 < len(compressed):
                b1, b2, b3 = compressed[i], compressed[i+1], compressed[i+2]
            elif i + 1 < len(compressed):
                b1, b2, b3 = compressed[i], compressed[i+1], 0
            else:
                b1, b2, b3 = compressed[i], 0, 0

            def encode6bit(b):
                if b < 10: return chr(48 + b)
                b -= 10
                if b < 26: return chr(65 + b)
                b -= 26
                if b < 26: return chr(97 + b)
                b -= 26
                return '-' if b == 0 else '_' if b == 1 else '?'

            res += encode6bit(b1 >> 2)
            res += encode6bit(((b1 & 0x3) << 4) | (b2 >> 4))
            res += encode6bit(((b2 & 0xF) << 2) | (b3 >> 6))
            res += encode6bit(b3 & 0x3F)
        return res

    try:
        encoded = encode_plantuml(code)
        url = f"https://www.plantuml.com/plantuml/png/{encoded}"

        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        response = urllib.request.urlopen(req, timeout=30)

        with open(output_path, 'wb') as f:
            f.write(response.read())

        return RenderResult(True, output_path, None, DiagramFormat.PLANTUML)

    except Exception as e:
        return RenderResult(False, None, str(e), DiagramFormat.PLANTUML)


def render_diagram(code: str, output_path: str, format_type: DiagramFormat = None) -> RenderResult:
    """
    Render diagram code to PNG.

    Args:
        code: The diagram code
        output_path: Path for output PNG
        format_type: The diagram format (auto-detected if None)

    Returns:
        RenderResult with rendering details
    """
    if format_type is None:
        validation = validate_diagram(code)
        format_type = validation.format

    renderers = {
        DiagramFormat.GRAPHVIZ: render_graphviz,
        DiagramFormat.PLANTUML: render_plantuml,
        DiagramFormat.MERMAID: render_mermaid,
    }

    return renderers[format_type](code, output_path)


# ============== Code Extraction Tools ==============

def extract_code_from_response(text: str, format_type: str = None) -> Optional[str]:
    """
    Extract diagram code from LLM response text.

    Args:
        text: The LLM response text
        format_type: Expected format ("graphviz", "mermaid", "plantuml")

    Returns:
        Extracted code or None
    """
    # Remove markdown code blocks
    patterns = [
        # Specific language blocks
        r'```(?:dot|graphviz)\s*\n(.*?)\n```',
        r'```(?:mermaid)\s*\n(.*?)\n```',
        r'```(?:plantuml|puml)\s*\n(.*?)\n```',
        # Generic blocks with graph content
        r'```\s*\n((?:di)?graph\s+\w+\s*\{.*?\})\s*```',
        r'```\s*\n(flowchart\s+\w+.*?)\n```',
        r'```\s*\n(@startuml.*?@enduml)\s*```',
    ]

    for pattern in patterns:
        matches = re.findall(pattern, text, re.DOTALL | re.IGNORECASE)
        if matches:
            return matches[0].strip()

    # Fallback: look for raw diagram code
    if format_type == 'graphviz' or re.search(r'(di)?graph\s+\w+\s*\{', text, re.IGNORECASE):
        match = re.search(r'((?:di)?graph\s+\w+\s*\{[\s\S]*\})', text, re.IGNORECASE)
        if match:
            return match.group(1).strip()

    if format_type == 'mermaid' or text.strip().lower().startswith(('flowchart', 'graph ')):
        # Return the whole text if it looks like mermaid
        cleaned = text.strip()
        cleaned = re.sub(r'^```\w*\s*\n?', '', cleaned)
        cleaned = re.sub(r'\n?```\s*$', '', cleaned)
        return cleaned if cleaned else None

    if format_type == 'plantuml' or '@startuml' in text.lower():
        match = re.search(r'(@startuml[\s\S]*?@enduml)', text, re.IGNORECASE)
        if match:
            return match.group(1).strip()

    return None

File: agents-vercel-sdk/src/tools.ts
/**
 * Diagram Agent Tools
 *
 * Reusable tools for diagram validation, rendering, and file operations.
 * TypeScript implementation using Vercel AI SDK patterns.
 */

import { tool } from "ai";
import { z } from "zod";
import { exec } from "child_process";
import { promisify } from "util";
import * as fs from "fs/promises";
import * as path from "path";
import * as os from "os";

const execAsync = promisify(exec);

// ============== Types ==============

export enum DiagramFormat {
  GRAPHVIZ = "graphviz",
  PLANTUML = "plantuml",
  MERMAID = "mermaid",
}

export interface ValidationResult {
  valid: boolean;
  format: DiagramFormat;
  errors: string[];
  code: string;
  suggestions: string[];
}

export interface RenderResult {
  success: boolean;
  outputPath: string | null;
  error: string | null;
  format: DiagramFormat;
}

// ============== Validation Functions ==============

export function validateGraphviz(code: string): ValidationResult {
  const errors: string[] = [];
  const suggestions: string[] = [];

  // Basic syntax checks
  if (!/(di)?graph\s+\w+\s*\{/i.test(code)) {
    errors.push(
      "No valid graph definition found (digraph Name { or graph Name {)"
    );
    suggestions.push("Start with: digraph MyDiagram {");
  }

  // Check brace balance
  const openBraces = (code.match(/\{/g) || []).length;
  const closeBraces = (code.match(/\}/g) || []).length;
  if (openBraces !== closeBraces) {
    errors.push(`Unbalanced braces: ${openBraces} '{' vs ${closeBraces} '}'`);
    suggestions.push("Ensure every '{' has a matching '}'");
  }

  // Check for common LLM mistakes
  if (/\.\w+\s*\[/.test(code)) {
    errors.push("Node names cannot start with dots (e.g., .env is invalid)");
    suggestions.push("Use underscores: env_file instead of .env");
  }

  if (/:\w+\s*->/.test(code)) {
    errors.push("Port syntax (:port) in node names may cause issues");
    suggestions.push("Define ports separately using record shapes");
  }

  return {
    valid: errors.length === 0,
    format: DiagramFormat.GRAPHVIZ,
    errors,
    code,
    suggestions,
  };
}

export function validateMermaid(code: string): ValidationResult {
  const errors: string[] = [];
  const suggestions: string[] = [];

  const codeStripped = code.trim().toLowerCase();

  // Check for valid start
  const validStarts = [
    "flowchart",
    "graph",
    "sequencediagram",
    "classdiagram",
    "erdiagram",
    "gantt",
    "pie",
    "statediagram",
  ];

  const hasValidStart = validStarts.some((s) => codeStripped.startsWith(s));
  if (!hasValidStart) {
    errors.push(
      "Must start with a valid diagram type (flowchart, graph, etc.)"
    );
    suggestions.push("Start with: flowchart LR or graph TD");
  }

  // Check bracket balance
  const openSquare = (code.match(/\[/g) || []).length;
  const closeSquare = (code.match(/\]/g) || []).length;
  if (openSquare !== closeSquare) {
    errors.push(
      `Unbalanced square brackets: ${openSquare} '[' vs ${closeSquare} ']'`
    );
    suggestions.push("Ensure every '[' has a matching ']'");
  }

  // Check subgraph/end balance
  const subgraphCount = (code.match(/\bsubgraph\b/gi) || []).length;
  const endCount = (code.match(/\bend\b/gi) || []).length;
  if (subgraphCount !== endCount) {
    errors.push(
      `Unbalanced subgraph/end: ${subgraphCount} subgraph vs ${endCount} end`
    );
    suggestions.push("Every 'subgraph' needs a closing 'end'");
  }

  return {
    valid: errors.length === 0,
    format: DiagramFormat.MERMAID,
    errors,
    code,
    suggestions,
  };
}

export function validatePlantuml(code: string): ValidationResult {
  const errors: string[] = [];
  const suggestions: string[] = [];

  const codeLower = code.toLowerCase();

  if (!codeLower.includes("@startuml")) {
    errors.push("Missing @startuml at the beginning");
    suggestions.push("Start with: @startuml");
  }

  if (!codeLower.includes("@enduml")) {
    errors.push("Missing @enduml at the end");
    suggestions.push("End with: @enduml");
  }

  if (codeLower.includes("@startuml") && codeLower.includes("@enduml")) {
    const startPos = codeLower.indexOf("@startuml");
    const endPos = codeLower.indexOf("@enduml");
    if (startPos > endPos) {
      errors.push("@startuml must come before @enduml");
    }
  }

  return {
    valid: errors.length === 0,
    format: DiagramFormat.PLANTUML,
    errors,
    code,
    suggestions,
  };
}

export function validateDiagram(
  code: string,
  formatHint?: string
): ValidationResult {
  code = code.trim();

  // Auto-detect format
  let detectedFormat: string;
  if (formatHint) {
    detectedFormat = formatHint.toLowerCase();
  } else if (code.toLowerCase().includes("@startuml")) {
    detectedFormat = "plantuml";
  } else if (
    code.toLowerCase().startsWith("flowchart") ||
    code.toLowerCase().startsWith("graph ")
  ) {
    detectedFormat = "mermaid";
  } else if (/(di)?graph\s+\w+\s*\{/i.test(code)) {
    detectedFormat = "graphviz";
  } else {
    detectedFormat = "graphviz";
  }

  const validators: Record<string, (code: string) => ValidationResult> = {
    graphviz: validateGraphviz,
    plantuml: validatePlantuml,
    mermaid: validateMermaid,
  };

  return validators[detectedFormat]?.(code) || validateGraphviz(code);
}

// ============== Rendering Functions ==============

export async function renderGraphviz(
  code: string,
  outputPath: string
): Promise<RenderResult> {
  try {
    const tmpFile = path.join(os.tmpdir(), `diagram_${Date.now()}.dot`);
    await fs.writeFile(tmpFile, code);

    await execAsync(`dot -Tpng -Gdpi=150 "${tmpFile}" -o "${outputPath}"`);
    await fs.unlink(tmpFile);

    return {
      success: true,
      outputPath,
      error: null,
      format: DiagramFormat.GRAPHVIZ,
    };
  } catch (error) {
    return {
      success: false,
      outputPath: null,
      error: error instanceof Error ? error.message : String(error),
      format: DiagramFormat.GRAPHVIZ,
    };
  }
}

export async function renderMermaid(
  code: string,
  outputPath: string
): Promise<RenderResult> {
  try {
    const encoded = Buffer.from(code).toString("base64url");
    const url = `https://mermaid.ink/img/${encoded}?type=png&bgColor=white`;

    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const buffer = await response.arrayBuffer();
    await fs.writeFile(outputPath, Buffer.from(buffer));

    return {
      success: true,
      outputPath,
      error: null,
      format: DiagramFormat.MERMAID,
    };
  } catch (error) {
    return {
      success: false,
      outputPath: null,
      error: error instanceof Error ? error.message : String(error),
      format: DiagramFormat.MERMAID,
    };
  }
}

export async function renderPlantuml(
  code: string,
  outputPath: string
): Promise<RenderResult> {
  try {
    // PlantUML encoding
    const deflate = await import("zlib");
    const compressed = deflate.deflateSync(code);
    const encoded = encodePlantUML(compressed.subarray(2, -4));
    const url = `https://www.plantuml.com/plantuml/png/${encoded}`;

    const response = await fetch(url, {
      headers: { "User-Agent": "Mozilla/5.0" },
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const buffer = await response.arrayBuffer();
    await fs.writeFile(outputPath, Buffer.from(buffer));

    return {
      success: true,
      outputPath,
      error: null,
      format: DiagramFormat.PLANTUML,
    };
  } catch (error) {
    return {
      success: false,
      outputPath: null,
      error: error instanceof Error ? error.message : String(error),
      format: DiagramFormat.PLANTUML,
    };
  }
}

function encodePlantUML(data: Buffer): string {
  const encode6bit = (b: number): string => {
    if (b < 10) return String.fromCharCode(48 + b);
    b -= 10;
    if (b < 26) return String.fromCharCode(65 + b);
    b -= 26;
    if (b < 26) return String.fromCharCode(97 + b);
    b -= 26;
    return b === 0 ? "-" : b === 1 ? "_" : "?";
  };

  let result = "";
  for (let i = 0; i < data.length; i += 3) {
    const b1 = data[i] || 0;
    const b2 = data[i + 1] || 0;
    const b3 = data[i + 2] || 0;
    result += encode6bit(b1 >> 2);
    result += encode6bit(((b1 & 0x3) << 4) | (b2 >> 4));
    result += encode6bit(((b2 & 0xf) << 2) | (b3 >> 6));
    result += encode6bit(b3 & 0x3f);
  }
  return result;
}

export async function renderDiagram(
  code: string,
  outputPath: string,
  formatType?: DiagramFormat
): Promise<RenderResult> {
  if (!formatType) {
    const validation = validateDiagram(code);
    formatType = validation.format;
  }

  const renderers: Record<
    DiagramFormat,
    (code: string, outputPath: string) => Promise<RenderResult>
  > = {
    [DiagramFormat.GRAPHVIZ]: renderGraphviz,
    [DiagramFormat.PLANTUML]: renderPlantuml,
    [DiagramFormat.MERMAID]: renderMermaid,
  };

  return renderers[formatType](code, outputPath);
}

// ============== Code Extraction ==============

export function extractCodeFromResponse(
  text: string,
  formatType?: string
): string | null {
  // Remove markdown code blocks
  const patterns = [
    /```(?:dot|graphviz)\s*\n([\s\S]*?)\n```/i,
    /```(?:mermaid)\s*\n([\s\S]*?)\n```/i,
    /```(?:plantuml|puml)\s*\n([\s\S]*?)\n```/i,
    /```\s*\n((di)?graph\s+\w+\s*\{[\s\S]*?\})\s*```/i,
    /```\s*\n(flowchart\s+\w+[\s\S]*?)\n```/i,
    /```\s*\n(@startuml[\s\S]*?@enduml)\s*```/i,
  ];

  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) {
      return match[1].trim();
    }
  }

  // Fallback: look for raw diagram code
  if (formatType === "graphviz" || /(di)?graph\s+\w+\s*\{/i.test(text)) {
    const match = text.match(/((?:di)?graph\s+\w+\s*\{[\s\S]*\})/i);
    if (match) return match[1].trim();
  }

  if (
    formatType === "mermaid" ||
    text.trim().toLowerCase().startsWith("flowchart") ||
    text.trim().toLowerCase().startsWith("graph ")
  ) {
    let cleaned = text.trim();
    cleaned = cleaned.replace(/^```\w*\s*\n?/, "");
    cleaned = cleaned.replace(/\n?```\s*$/, "");
    return cleaned || null;
  }

  if (formatType === "plantuml" || text.toLowerCase().includes("@startuml")) {
    const match = text.match(/(@startuml[\s\S]*?@enduml)/i);
    if (match) return match[1].trim();
  }

  return null;
}

// ============== AI SDK Tool Definitions ==============

export const validateDiagramTool = tool({
  description:
    "Validate diagram code and check for syntax errors. Returns validation result with errors and suggestions.",
  parameters: z.object({
    code: z.string().describe("The diagram code to validate"),
    formatHint: z
      .enum(["graphviz", "mermaid", "plantuml"])
      .optional()
      .describe("Optional format hint"),
  }),
  execute: async ({ code, formatHint }) => {
    const result = validateDiagram(code, formatHint);
    return {
      valid: result.valid,
      format: result.format,
      errors: result.errors,
      suggestions: result.suggestions,
    };
  },
});

export const renderDiagramTool = tool({
  description: "Render diagram code to PNG file.",
  parameters: z.object({
    code: z.string().describe("The diagram code to render"),
    outputPath: z.string().describe("Path for output PNG file"),
    formatType: z
      .enum(["graphviz", "mermaid", "plantuml"])
      .optional()
      .describe("Diagram format"),
  }),
  execute: async ({ code, outputPath, formatType }) => {
    const fmt = formatType
      ? (DiagramFormat[
          formatType.toUpperCase() as keyof typeof DiagramFormat
        ] as DiagramFormat)
      : undefined;
    const result = await renderDiagram(code, outputPath, fmt);
    return {
      success: result.success,
      outputPath: result.outputPath,
      error: result.error,
      format: result.format,
    };
  },
});

export const extractCodeTool = tool({
  description: "Extract diagram code from LLM response text.",
  parameters: z.object({
    responseText: z.string().describe("The LLM response containing diagram code"),
    expectedFormat: z
      .enum(["graphviz", "mermaid", "plantuml"])
      .optional()
      .describe("Expected format"),
  }),
  execute: async ({ responseText, expectedFormat }) => {
    const code = extractCodeFromResponse(responseText, expectedFormat);
    return code || "ERROR: Could not extract diagram code from response";
  },
});
